File: 01-week-1--starter-algorithms/03-day-4--selection-sort/ruby/selection_sort.rb
def selection_sort(arr)
  output = []

  while arr.length > 0 do
    min = arr.min
    i = arr.index(min)
    output << min
    arr.delete_at(i)
  end

  output
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort([3, -1, 5, 2])

  puts

  # Don't forget to add your own!
  puts "Expecting: []"
  print "=> "
  print selection_sort([])

  puts

  puts "Expecting: [-1, 2, 5, 5]"
  print "=> "
  print selection_sort([5, -1, 5, 2])

  puts

  # BENCHMARK HERE, you can print the average runtime
  start_time = Time.now
  long_input = []

  100.times { long_input << rand }

  1000.times do
    selection_sort([3, 7, 2])
    selection_sort(long_input)
  end

  end_time = Time.now

  puts "Average runtime: #{(end_time - start_time) / 2000} millisecond(s)"
end

# Please add your pseudocode to this file
# - Use Math.min() to find the lowest value in input array
# - Remove that value and push it onto an output array
# - Continue the process until input array empty
# - Return output array

# And a written explanation of your solution
# Runs a loop once for every element of the input array in which the lowest value is removed from the input array and pushed onto the output array, which is returned.

File: 01-week-1--starter-algorithms/03-day-4--selection-sort/javascript/selection_sort.js
function selectionSort(arr) {
  let output = [];
  let length = arr.length;

  for (let i=0; i < length; i++) {
    let lowest = Math.min(...arr);
    let index = arr.indexOf(lowest);
    arr.splice(index, 1);
    output.push(lowest);
  }

  return output;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSort([3, -1, 5, 2]));

  console.log("");

  console.log("Expecting: [-1, 3, 5, 5, 6]");
  console.log("=>", selectionSort([3, -1, 5, 6, 5]));

  console.log("");

  console.log("Expecting: [-6, -5, -5, -3, -1]");
  console.log("=>", selectionSort([-3, -1, -5, -6, -5]));

  console.log("");

  // BENCHMARK HERE, and print the average runtime
  const startTime = Date.now();
  let shortInput = [];
  let longInput = [];

  for (let i = 0; i < 3; ++i) {
    shortInput.push(Math.random());
    longInput.push(Math.random());
  }

  for (let i = 0; i < 100; ++i) {
    shortInput.push(Math.random());
    longInput.push(Math.random());
  }

  for (let i = 0; i < 1000; i++) {
    selectionSort(shortInput);
    selectionSort(longInput);
  }

  let endTime = Date.now();
  let runtime = endTime - startTime;

  console.log(`Runtime: ${runtime} millisecond(s)`);
  console.log(`Average runtime: ${runtime / 2000} millisecond(s)`);
}

module.exports = selectionSort;

// Please add your pseudocode to this file
/*
- Use Math.min() to find the lowest value in input array
- Remove that value and push it onto an output array
- Continue the process for each element in input array
- Return output array
*/

// And a written explanation of your solution
/*
Runs a loop once for every element of the input array in which the lowest value is removed from the input array and pushed onto the output array, which is returned.
*/
